pageserver/src/tenant/timeline.rs

```diff
@@ -4155,6 +4161,171 @@ impl Timeline {
+                                //
+                                // Unfortunately we cannot do this for the main fork, or for
+                                // any metadata keys, keys, as that would lead to actual data
+                                // loss.
+                                if is_rel_fsm_block_key(img_key) || is_rel_vm_block_key(img_key) {
+                                    warn!("could not reconstruct FSM or VM key {img_key}, filling with zeros: {err:?}");
+                                    ZERO_PAGE.clone()
+                                } else {
+                                    return Err(CreateImageLayersError::PageReconstructError(err));
+                                }
+                            }
+                        };
+
+                        // Write all the keys we just read into our new image layer.
+                        image_layer_writer.put_image(img_key, img, ctx).await?;
+                        wrote_keys = true;
+                    }
+                }
+            }
+        }
+
+        if wrote_keys {
+            // Normal path: we have written some data into the new image layer for this
+            // partition, so flush it to disk.
+            *start = img_range.end;
+            let image_layer = image_layer_writer.finish(self, ctx).await?;
+            Ok(Some(image_layer))
+        } else {
+            // Special case: the image layer may be empty if this is a sharded tenant and the
+            // partition does not cover any keys owned by this shard.  In this case, to ensure
+            // we don't leave gaps between image layers, leave `start` where it is, so that the next
+            // layer we write will cover the key range that we just scanned.
+            tracing::debug!("no data in range {}-{}", img_range.start, img_range.end);
+            Ok(None)
+        }
+    }
+
+    /// Create an image layer for metadata keys. This function produces one image layer for all metadata
+    /// keys for now. Because metadata keys cannot exceed basebackup size limit, the image layer for it
+    /// would not be too large to fit in a single image layer.
+    #[allow(clippy::too_many_arguments)]
+    async fn create_image_layers_for_metadata_keys(
+        self: &Arc<Self>,
+        partition: &KeySpace,
+        mut image_layer_writer: ImageLayerWriter,
+        lsn: Lsn,
+        ctx: &RequestContext,
+        img_range: Range<Key>,
+        start: &mut Key,
+        mode: ImageLayerCreationMode,
+    ) -> Result<Option<ResidentLayer>, CreateImageLayersError> {
+        // Metadata keys image layer creation.
+        let delta_file_accessed_begin = ctx
+            .vectored_access_delta_file_cnt
+            .load(AtomicOrdering::SeqCst);
+        let data = self
+            .get_vectored_impl(
+                partition.clone(),
+                lsn,
+                ValuesReconstructState::default(),
+                ctx,
+            )
+            .await?;
```

Comment:
Why not `Timeline::scan`? At least that does some throttling...

Reply:
I thought compaction (or create image layer) itself is already throttled and we don't need to throttle it again?

Reply:
It's throttled in the sense that we limit the number of concurrent compactions that run at once. Under the assumption that metadata key space is very sparse not throttling is not a big deal, but then again compaction for rel blocks gets throttled correctly.

Reply:
Anyways, the difference between one semaphore unit and 0 semaphore units is kinda insignificant

Reply:
Another reason I want to keep using `get_vectored_impl` is that we need to get `ValuesReconstructState` back... where `scan` does not expose this to the user of the interface.

Reply:
Yeah, my comments were contradictory in that regard. Leaving it up to you to decide if we should worry about throttling for now